### PR TITLE
[SPARK-14773][SPARK-15179][SQL] Fix SQL building and enable Hive tests

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?

This PR fixes SQL building for predicate subqueries and correlated scalar subqueries. It also enables most Hive subquery tests.
## How was this patch tested?

Enabled new tests in HiveComparisionSuite.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @rxin @davies @liancheng

**Comment by hvanhovell** [comment]
I'll probably put this on a diet. I don't think we need the actual explain outputs...

**Comment by davies** [comment]
@hvanhovell We should remove those explain outputs (remove the file or truncate it to empty file)

**Comment by hvanhovell** [comment]
@davies done. It saves ~3KLOC.

**Comment by davies** [comment]
LGTM

**Comment by yhuai** [comment]
@hvanhovell How about we also add the regression tests to `LogicalPlanToSQLSuite` (in a separate PR)? It will be good to reduce the dependency on HiveCompatibilitySuite.

**Comment by hvanhovell** [comment]
@yhuai I didn't know we had that test. I have created https://issues.apache.org/jira/browse/SPARK-15232 to track this.

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala [hunk 1]
```diff
@@ -61,6 +62,15 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     // Ensures that the plans generation use metastore relation and not OrcRelation
     // Was done because SqlBuilder does not work with plans having logical relation
     TestHive.setConf(HiveUtils.CONVERT_METASTORE_ORC, false)
+
```
Comment (hvanhovell) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: I'll remove this...

**Comment by hvanhovell** [comment]
@yhuai is this one OK to merge?

**Comment by davies** [comment]
Merging this into master and 2.0, thanks!

**Pull request closed without merging**